### PR TITLE
Feature: Global authentication context

### LINKED ISSUE #54: Global authentication context
**Describe the task that needs to be done.**

A global authentication context needs to be developed for the application, whose responsibility is to manage the user's JWT token. This context will primarily be used by the `Route` middleware described in #16.

**Describe how a solution to your proposed task might look like (and any alternatives considered).**

We can create an `AuthContext` using `React.createContext({})`, and then wrap any data operations and state in a component called `AuthContextProvider`. For managing the JWT state, there are two approaches that can be used.

1. Use a package called `use-persisted-state` to store the JWT in the local storage.
2. Use the normal state in React to store the JWT. The only disadvantage here is that the user will need to log in again if they refresh the application because the state will be reset.

**Notes**

- The high-level detail of this approach was agreed by everyone in the frontend meeting on 10th March.

### PR DESCRIPTION
**Pull request by shreytailor**
## Description

The intent of this pull request is to implement a context provider that manages the JWT token within the application. The `<App/>` component is wrapped around this `<AuthContextProvider/>` so that all components within our application can get access to it.

When the application starts up, the context checks whether there is a key in the browser's local storage called 'jwt'. If not, it initializes the key with a value of an empty string. The context also provides a method called `setJwt(string)` which is used to set the JWT to the localStorage when the user signs in successfully.

A `logout(string)` method is also present which basically clears the JWT from the local storage. When a JWT is set to the context, it performs an initial check with the `/verify` endpoint whether it's valid.

Closes #54

## How has this been tested?

Since the component is tightly coupled with dependencies related to the Browser API (e.g. localStorage) and backend API calls, no further unit tests have been written. However, the component was manually tested against a mock API producing either a success/failure code depending on the value of JWT provided.

Note: once the login page is up and running, maybe we can add integration tests to check if it works correctly.

## Screenshots

N/A

## Checklist
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: src/index.jsx [hunk 1]
```diff
@@ -1,12 +1,15 @@
...
+import { AuthContextProvider } from './context/authContext';
 import { ServiceWorkerProvider } from './context/serviceWorkerContext';
 
 ReactDOM.render(
   <React.StrictMode>
     <ServiceWorkerProvider>
-      <App />
+      <AuthContextProvider>
+        <App />
+      </AuthContextProvider>
```
Comment (scorpionknifes) [inline src/index.jsx:12]: thought on moving this provider to `App.jsx`?

**Review by scorpionknifes — APPROVED** [review]
LGTM

File: src/index.jsx (hunk 1 above)
Comment (shreytailor) [reply src/index.jsx:12]: So you mean wrapping everything inside `App.jsx` with `AuthContextProvider` rather than here?

**Comment by shreytailor** [comment]
> LGTM, do we need unit tests?

Do you have any idea on how we can unit test this?

I tried to change the prototype of `localStorage` methods globally and tried testing it, but it didn't work as intended.

File: src/index.jsx (hunk 1 above)
Comment (scorpionknifes) [reply src/index.jsx:12]: yeah

**Review by scorpionknifes — APPROVED** [review]

**Pull request merged**